pkg/controllers/v1alpha2/mpi_job_controller.go

```diff
@@ -0,0 +1,1158 @@
+		// work queue.
+		if key, ok = obj.(string); !ok {
+			// As the item in the work queue is actually invalid, we call
+			// Forget here else we'd go into a loop of attempting to
+			// process a work item that is invalid.
+			c.queue.Forget(obj)
+			runtime.HandleError(fmt.Errorf("expected string in workqueue but got %#v", obj))
+			return nil
+		}
+		// Run the syncHandler, passing it the namespace/name string of the
+		// MPIJob resource to be synced.
+		if err := c.syncHandler(key); err != nil {
+			return fmt.Errorf("error syncing '%s': %s", key, err.Error())
+		}
+		// Finally, if no error occurs we Forget this item so it does not
+		// get queued again until another change happens.
+		c.queue.Forget(obj)
+		glog.Infof("Successfully synced '%s'", key)
+		return nil
+	}(obj)
+
+	if err != nil {
+		runtime.HandleError(err)
+		return true
+	}
+
+	return true
+}
+
+// syncHandler compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the MPIJob resource
+// with the current status of the resource.
+func (c *MPIJobController) syncHandler(key string) error {
+	// Convert the namespace/name string into a distinct namespace and name.
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	// Get the MPIJob with this namespace/name.
+	mpiJob, err := c.mpiJobLister.MPIJobs(namespace).Get(name)
+	// The MPIJob may no longer exist, in which case we stop processing.
+	if errors.IsNotFound(err) {
+		runtime.HandleError(fmt.Errorf("mpi job '%s' in work queue no longer exists", key))
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	// Get the launcher Job for this MPIJob.
+	launcher, err := c.getLauncherJob(mpiJob)
+	if err != nil {
+		return err
+	}
+	// We're done if the launcher either succeeded or failed.
+	done := launcher != nil && (launcher.Status.Succeeded == 1 || launcher.Status.Failed == 1)
+
+	if err != nil {
+		runtime.HandleError(err)
+		return nil
+	}
+
+	workerSpec := mpiJob.Spec.MPIReplicaSpecs[kubeflow.MPIReplicaTypeWorker]
+	workerReplicas := int(*workerSpec.Replicas)
+
+	if !done {
+		// Get the ConfigMap for this MPIJob.
+		if config, err := c.getOrCreateConfigMap(mpiJob, workerReplicas); config == nil || err != nil {
+			return err
+		}
+
+		// Get the launcher ServiceAccount for this MPIJob.
+		if sa, err := c.getOrCreateLauncherServiceAccount(mpiJob); sa == nil || err != nil {
+			return err
+		}
+
+		// Get the launcher Role for this MPIJob.
+		if r, err := c.getOrCreateLauncherRole(mpiJob, workerReplicas); r == nil || err != nil {
+			return err
+		}
+
+		// Get the launcher RoleBinding for this MPIJob.
+		if rb, err := c.getLauncherRoleBinding(mpiJob); rb == nil || err != nil {
+			return err
+		}
+
+		// Get the PDB for this MPIJob
+		if c.enableGangScheduling {
+			if pdb, err := c.getOrCreatePDB(mpiJob, workerReplicas); pdb == nil || err != nil {
```

Comment:
should move to podgroup here :) pdb has deprecated in kube-batch

Reply:
For sure yes but it's probably better done in a separate PR. Contribution is welcomed!

Reply:
Okey,so we should push this pr merged first

Reply:
Just updated the PR and commented. Feel free to help continue working on this if you are interested.

Reply:
sure, I will interested working on this :)